pkg/datastore/deploymentchainstore.go

```diff
@@ -71,10 +74,24 @@ var (
 				node.DeploymentRef.Status = status
 				node.DeploymentRef.StatusReason = reason
 				updated = true
+				break
 			}
+
 			if !updated {
 				return fmt.Errorf("unable to find the right node in chain to assign deployment to")
 			}
+
+			// If the block is already finished, keep its finished status.
+			if block.IsCompleted() {
+				return nil
+			}
+
+			// Update block status based on its state after update latest submitted deployment status.
+			block.Status = block.DesiredStatus()
+			if block.IsCompleted() {
+				block.CompletedAt = time.Now().Unix()
+			}
+
```

Comment:
I think we should not write too much logic in this datastore package. Instead of that, moving all of them to model package for more testable.

For example, instead of defining `DeploymentChainNodeDeploymentStatusUpdater` function in this package, we can have below one in the model package to use in API.

``` golang
func(c *DeploymentChain) UpdateDeploymentStatus(block unit32, deploymentID string, status DeploymentStatus, reason string) error {
}
```

Reply:
That looks like a side-effect containing function and I don't think it's better than the current implementation for testing, since I have written tests for this updater function as well, how do you think about that.

Reply:
What is the side-effect are you talking about?
By moving to the model package, that `DeploymentChainNodeDeploymentStatusUpdater` will be no longer needed, we can use the `UpdateDeploymentStatus` function directly in the API.

https://github.com/pipe-cd/pipe/blob/master/pkg/app/api/grpcapi/piped_api.go#L554

One good point we can have is that all logic code around the model object is stayed inside that package, only it.
This datastore package could be quite simple.

Reply:
I understand your thought, but lets say the side-effect I'm talking about is
> dc: deploymentChainModel => dc.UpdateItselfFunction() => dc' => write dc' to the database at the key of dc obj

I think it would be better to be like
> dc: deploymentChainModel => updater function(dc, updateMethod) => datastore uses updater to update model dc

Reply:
Sorry, I don't get that explanation clearly.
Let me write explicitly what I am thinking here so you can compare them.

### Here is the current implementation

- In the datastore package

``` golang
  DeploymentChainNodeDeploymentStatusUpdater = func(blockIndex uint32, deploymentID string, status model.DeploymentStatus, reason string) func(*model.DeploymentChain) error {
    return func(dc *model.DeploymentChain) error {
      // logic code
    }
  }
```

- In the API package
https://github.com/pipe-cd/pipe/blob/master/pkg/app/api/grpcapi/piped_api.go#L554

``` golang
  dcUpdater := datastore.DeploymentChainNodeDeploymentStatusUpdater(blockIndex, deploymentID, status, reason)
  if err := a.deploymentChainStore.UpdateDeploymentChain(ctx, deploymentChainID, dcUpdater); err != nil {
    return err
  }
```

### Here is the one I am talking about

- In the model package

``` golang
  func(c *DeploymentChain) UpdateDeploymentStatus(block unit32, deploymentID string, status DeploymentStatus, reason string) error {
  }
```

- In the API package
https://github.com/pipe-cd/pipe/blob/master/pkg/app/api/grpcapi/piped_api.go#L554

``` golang
  err := a.deploymentChainStore.UpdateDeploymentChain(ctx, deploymentChainID, func(c *model.DeploymentChain) error {
    return c.UpdateDeploymentStatus(blockIndex, deploymentID, status, reason)
  }
```

So where is the side-effect?

Reply:
Thanks for the explanation 👍 
What I mean is the `c.UpdateDeploymentStatus(blockIndex, deploymentID, status, reason)` is side-effect containing function, since the `c` object itself call its method and becomes `c'`. I usually avoid making methods that way to avoid misleading, since for instance if this `c.UpdateDeploymentStatus(...)` is called somewhere else (not inside the updater function), it will change the values of that `c` object itself, and all code after that line is tricky. The current updater is also a side-effect containing function, but since it's named updater and we only use it in UpdateDeploymentChain, it's understandable and we can't use it anywhere else, that's my point.

Reply:
I think `side-effect` is when it modifies something out of its scope.
In our case, that `Update` function is only updating its fields so that is a normal case and could not be called as causing `side-effect`.

We are not restricting the model package to have only `read-only` functions, but any model can have `write` functions to update its fields.

Reply:
> I think side-effect is when it modifies something out of its scope.

You're right, I used the wrong word for that case 🙏 
Besides, I have checked our current models once, bellow is the result
```console
╰─ ag --go --ignore "*.pb.*" "Update" pkg/model 
pkg/model/deployment.go
73:// CanUpdateDeploymentStatus checks whether the deployment can transit to the given status.
74:func CanUpdateDeploymentStatus(cur, next DeploymentStatus) bool {
99:// CanUpdateStageStatus checks whether the stage can transit to the given status.
100:func CanUpdateStageStatus(cur, next StageStatus) bool {

pkg/model/project.go
67:// Update updates ProjectStaticUser with given data.
68:func (p *ProjectStaticUser) Update(username, password string) error {
108:// Update updates ProjectSSOConfig with given data.
109:func (p *ProjectSSOConfig) Update(sso *ProjectSSOConfig) error {
115:            if err := p.Github.Update(sso.Github); err != nil {
168:// Update updates ProjectSSOConfig with given data.
169:func (p *ProjectSSOConfig_GitHub) Update(input *ProjectSSOConfig_GitHub) error {

pkg/model/project_test.go
79:func TestUpdateProjectStaticUser(t *testing.T) {
99:                     p.Update(tc.username, tc.password)
107:func TestUpdateProjectSSOConfig(t *testing.T) {
141:                    p.Update(tc.sso)
```
As you can see, the only place we explicitly use update/write functions to change object itself values is the project model (I'm not really like that), and besides, the `deployment` model has functions like `CanUpdateDeploymentStatus` and `CanUpdateStageStatus` which be used to determine updating logic in updater functions in its datastore updater (just as I do with the `DesiredStatus` and deploymentChainStore updater here). Tbh, I think the model objects should be kept immutable (just same as the word `read-only`?) as much as possible, all update/change the model objects itself should be placed in the place we can not use after we make that change (API layer before we return back to the client or database layer before we store in the datastore in updater function). Go does not strict that way but feel like it is safer, since once we make a function like `UpdateDeploymentStatus` it will: update the status of the specified deployment, update the status of container block if necessary and update CompletedAt value in case the block completed, that's too much for the name `UpdateDeploymentStatus`, IMO. And we have no way to ensure we will not accidentally use that function with the wrong purpose.

> any model can have write functions to update its fields.

Yes, I know that, since the model's fields are public, so basically we can explicitly change the value directly (not even need a setter or something like that), but mostly it language specified or generated code, so it should be counted as unavoidable. What we can do is try to reduce the risk. How do you think?

Reply:
On the second thought, I think the good thing about adding `UpdateDeploymentStatus` to the model is that we can have the deployment chain object after it's changed without re-fetch it from the datastore; that's both nice and not nice at the same time 😂 Since we can have logic which based on something that not actually be consisted and unsafe. I think I will not go further with this discussion since it's not the main thing in the PR, so if we can find out some names that avoid us from misleading on using this model function in the future, I will be okay with that. The current name is `UpdateDeploymentStatus`, but I think it should be kind of `UpdateDeploymentStatusAndChangeStatus` (since it's ChainBlock model function, the last Status means to ChainBlock.Status), but it's hard to read.

Reply:
OK. Let's keep the current implementation since that thing is out of the scope of this PR.

Reply:
Thank you 🙏